src/libstd/io/stdio.rs

```diff
@@ -155,6 +143,121 @@ fn handle_ebadf<T>(r: io::Result<T>, default: T) -> io::Result<T> {
+        let cap = stdio::STDIN_BUF_SIZE;
+        StdioReader::Buffered(BufReader::with_capacity(cap, HandleEbadf(reader)))
+    }
+    fn new_fake() -> StdioReader<R> {
+        StdioReader::Fake
+    }
+}
+
+impl<R: Read> Read for StdioReader<R> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match self {
+            StdioReader::Buffered(i) => i.read(buf),
+            StdioReader::Fake => Ok(0),
+        }
+    }
+    fn read_vectored(&mut self, bufs: &mut [io::IoSliceMut<'_>]) -> io::Result<usize> {
+        match self {
+            StdioReader::Buffered(i) => i.read_vectored(bufs),
+            StdioReader::Fake => Ok(0),
+        }
+    }
+    #[inline]
+    unsafe fn initializer(&self) -> Initializer {
+        match self {
+            StdioReader::Buffered(i) => i.initializer(),
+            StdioReader::Fake => Initializer::nop(),
+        }
+    }
+}
+
+impl<R: Read> BufRead for StdioReader<R> {
+    fn fill_buf(&mut self) -> io::Result<&[u8]> {
+        match self {
+            StdioReader::Buffered(i) => i.fill_buf(),
+            StdioReader::Fake => Ok(&[]),
+        }
+    }
+    fn consume(&mut self, amt: usize) {
+        match self {
+            StdioReader::Buffered(i) => i.consume(amt),
+            StdioReader::Fake => (),
+        }
+    }
+}
+
+/// A writer that can either be line-buffered, buffered, unbuffered or fake.
+///
+/// If it is fake, all outputs just succeed and are dropped silently.
+enum StdioWriter<W: Write> {
+    LineBuffered(LineWriter<HandleEbadf<W>>),
+    Buffered(BufWriter<HandleEbadf<W>>),
+    Unbuffered(HandleEbadf<W>),
+    Fake,
+}
+
+impl<W: Write> StdioWriter<W> {
+    fn new(writer: W, kind: StdioBufferKind) -> StdioWriter<W> {
+        let writer = HandleEbadf(writer);
+        match kind {
+            StdioBufferKind::LineBuffered =>
+                StdioWriter::LineBuffered(LineWriter::new(writer)),
+            StdioBufferKind::Buffered =>
+                StdioWriter::Buffered(BufWriter::new(writer)),
+            StdioBufferKind::Unbuffered =>
+                StdioWriter::Unbuffered(writer),
+        }
+    }
+    fn new_fake() -> StdioWriter<W> {
+        StdioWriter::Fake
+    }
+}
+
+impl<W: Write> Write for StdioWriter<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match self {
+            StdioWriter::LineBuffered(i) => i.write(buf),
+            StdioWriter::Buffered(i) => i.write(buf),
+            StdioWriter::Unbuffered(i) => i.write(buf),
+            StdioWriter::Fake => Ok(buf.len()),
+        }
+    }
+    fn write_vectored(&mut self, bufs: &[io::IoSlice<'_>]) -> io::Result<usize> {
+        match self {
+            StdioWriter::LineBuffered(i) => i.write_vectored(bufs),
+            StdioWriter::Buffered(i) => i.write_vectored(bufs),
+            StdioWriter::Unbuffered(i) => i.write_vectored(bufs),
+            StdioWriter::Fake => Ok(bufs.iter().map(|b| b.len()).sum()),
```

Comment:
This sum could probably wrap around, seems like it should saturate instead (not sure what's reasonable), or return the first non-empty len.